mypy/checker.py

```diff
@@ -3286,7 +3317,10 @@ def check_assignment(
                             lvalue_type = make_optional_type(lvalue_type)
                             self.set_inferred_type(lvalue.node, lvalue, lvalue_type)
```

Comment:
IIUC this is old special-casing for inferring optional types from assignments in if statements. Shouldn't we make this conditional on new redefinition flag being `False`?

Reply:
Yeah, it seems that we can disable this logic when using the new redefinition flag.